WebJobs Analyzer: Function name validation

kashimiz opened the pull request:
Addresses #8258 

Related: https://github.com/Azure/azure-webjobs-sdk/pull/2842

Requires release of updated WebJobs SDK to build.

kashimiz committed:
Added function name validation analyzer. (#8258)

brettsam commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+        Dictionary<string, Assembly> _mapRef = new Dictionary<string, Assembly>();
+
+        const string WebJobsAssemblyName = "Microsoft.Azure.WebJobs";
+        const string WebJobsHostAssemblyName = "Microsoft.Azure.WebJobs.Host";
```
can you explicitly mark all these fields as `private`?

kashimiz replied:
I have removed the entire file. Will make these updates when I put it back.

brettsam commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+                _mapRef[WebJobsAssemblyName] = typeof(Microsoft.Azure.WebJobs.FunctionNameAttribute).Assembly;
+                _mapRef[WebJobsHostAssemblyName] = typeof(Microsoft.Azure.WebJobs.JobHost).Assembly;
+
+                // JSON.Net?
```
remove this comment?

kashimiz replied:
I have removed the entire file. Will make these updates when I put it back.

liliankasem commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+        // Assembly Display Name --> loaded Assembly object
+        Dictionary<string, Assembly> _mapRef = new Dictionary<string, Assembly>();
+
+        const string WebJobsAssemblyName = "Microsoft.Azure.WebJobs";
```
Do we want to put these in a constants file?

kashimiz replied:
Thanks for all the comments! This is a pretty straight port of code @MikeStall wrote in [azure/azure-webjobs-sdk ](https://github.com/azure/azure-webjobs-sdk/), so compliments go to him. :) (I added the tests.)

Good catch here; I'll move this into Constants.

liliankasem commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+                    string asmName = new AssemblyName(assembly.FullName).Name;
+                    _mapRef[asmName] = assembly;
+
+                    var test = assembly.GetCustomAttributes<WebJobsStartupAttribute>().Select(a => a.WebJobsStartupType);
```
Not sure that I understand why the variable is named test? It looks like you're selecting a startup type? I might be missing some context

kashimiz replied:
I think it means "I'm testing to see if this assembly has a `WebJobsStartupAttribute`" but I agree the name is a little indirect. I'll change it to `startupAttr`.

liliankasem commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+
+        private Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
+        {
+            var an = new AssemblyName(args.Name);
```
nit: I think it might be easier to follow the code with the full variable name
```suggestion
            var assemblyName = new AssemblyName(args.Name);
```

kashimiz replied:
Agreed here; changing.

liliankasem commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+            var an = new AssemblyName(args.Name);
+            var context = args.RequestingAssembly;
+
+            Assembly asm2;
```
what does asm2 stand for?

kashimiz replied:
I think it stands for "assembly2". There's not really an assembly1, so just changing this to `assembly` and changing other `asm[#]` instances.

liliankasem commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+
+                    if (string.Equals(an2.FullName, an.FullName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        var a = Assembly.LoadFrom(path);
```
similar nit here around using meaningful variable names i.e. `loadedAssembly`

kashimiz replied:
Changed.

liliankasem commented on src/WebJobs.Script.Analyzers/CompilationWebJobsStartuptypeLocator.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Microsoft.Azure.Functions.Analyzers
+{
+    public class CompilationWebJobsStartupTypeLocator : IWebJobsStartupTypeLocator
```
Mind adding a class comment here?

kashimiz replied:
Added!

kashimiz replied:
Added!

liliankasem commented on src/WebJobs.Script.Analyzers/Constants.cs:
```diff
@@ -14,6 +14,7 @@ internal static class DiagnosticsCategories
         {
             public const string Reliability = "Reliability";
             public const string Usage = "Usage";
+            public const string WebJobs = "WebJobs";    // TODO: Too vague?
```
What's the usage of this and what are you trying to represent? Context might help me answer that "todo" question :D

kashimiz replied:
This is a fun discussion opportunity. From what I've gleaned, we can define categories for our analyzers. Here's an example from Roslyn: https://github.com/dotnet/roslyn-analyzers/blob/main/src/Utilities/Compiler/DiagnosticCategory.cs

You can use categories to configure settings, ex. severity, suppression, etc: https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2022#set-rule-severity-of-multiple-analyzer-rules-at-once-in-an-editorconfig-file

Since we only have a few analyzers right now, I'm not sure how we want to divide them + future ones up. Since these analyzers operate on the WebJobs binding model, so not fully "Functions", I went with `WebJobs`, but also considered `Binding` as a possible name. Love to hear your thoughts here! + @fabiocav

liliankasem replied:
Ahh I see, would `WebJobsBindings` be a good category name?

kashimiz replied:
That'll do!

liliankasem commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(DiagnosticDescriptors.IllegalFunctionName); } }
+
+    public static void VerifyWebJobsLoaded() 
```
```suggestion
        public static void VerifyWebJobsLoaded() 
```

nit: misaligned

kashimiz replied:
Whoops! Thanks for the catch.

liliankasem commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+
+    public static void VerifyWebJobsLoaded() 
+        {
+            var x = new JobHost(new OptionsWrapper<JobHostOptions>(new JobHostOptions()), null);
```
nit?
```suggestion
            var jobHost = new JobHost(new OptionsWrapper<JobHostOptions>(new JobHostOptions()), null);
```

kashimiz replied:
Better name. Thanks!

liliankasem commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+                var val = semantics.GetConstantValue(arg.Expression);
+                if (!val.HasValue)
+                {
+                return null;
```
I think you can eliminate this return if you inverse the if to check for a value then return `return val.Value as string;` within the if condition, unless you're trying to not loop through everything and want to return early?

kshyju replied:
I think we are trying to check only the first argument. IMHO, explicitly reading the first argument might be a bit more readable. Something like this.
````
if (attributeSyntax.ArgumentList.Arguments.Count == 0)
{
    return null; 
}

var firstArgument = attributeSyntax.ArgumentList.Arguments[0];
var val = semantics.GetConstantValue(firstArgument.Expression);

return val.Value as string;
````

kashimiz replied:
Shyju is correct; the loop will hit a return statement on its first iteration. I'm not sure why it was written the way it was, but his suggested implementation is more straightforward. I'll make the substitution.

liliankasem commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+        {
+            foreach (var attrListSyntax in methodDeclarationSyntax.AttributeLists)
+            {
+                foreach (AttributeSyntax attributeSyntax in attrListSyntax.Attributes)
```
The double foreach scares be a bit perf wise, but I'm not sure if there is a better way without a data structure change? @brettsam ?

kshyju replied:
The inner foreach won't be executed for any method declarations which does not have attribute decorations.

liliankasem reviewed (commented):
I don't have a lot of context around assemblies so probably not the right person to sign off on this, but this looks great! Nice work, and I think you did a great job commenting your code which I appreciate a lot. Overall, just a few nits and questions :)

madelinegordon commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            _tooling = AssemblyCache.Instance.Tooling;
+
+            // cast to PortableExecutableReference which has a file path
+            var x1 = compilation.References.OfType<PortableExecutableReference>().ToArray();
```
nit: Could "x1" be made more descriptive?

kashimiz replied:
Very much so. Changed to `references`

madelinegordon commented on test/WebJobs.Script.Tests.Analyzers/WebJobsAttributeAnalyzerTests.cs:
```diff
@@ -0,0 +1,88 @@
+        [Fact]
+        public async Task ValidFunctionName_NoDiagnostic()
+        {
+            string testCode = @"
```
nit: Could this be extracted to a separate method like GetTestCode(string functionName), to make the test more readable?

kashimiz replied:
Yes; this approach worked when there were only a few tests, but now it's getting unwieldy. There are some other possible approaches, such as loading the code from .cs files in the test project. I'd like to flag this as a near-future enhancement (before things get too out of hand) and make those changes in a separate PR.

soninaren commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+                // By convention, restrict based on filenames.
+                var filename = Path.GetFileName(path);
+                // TODO: I assume this is so general to avoid missing custom extensions, but can it be tightened up for performance?
+                if (!filename.ToLowerInvariant().Contains("extension"))
```
`DurableTask.SqlServer.AzureFunctions.dll` does not have extension in the name. We may have to update this to include durable assemblies

soninaren replied:
Not sure if it is possible to read assemblyNames from extensions.json

kashimiz replied:
This logic is looking for `IWebJobsStartup` implementations to pull the relevant binding rules and validation attributes. The Durable Functions bindings are defined in `Microsoft.Azure.WebJobs.Extensions.DurableTask`, so I think skipping `DurableTask.SqlServer.AzureFunctions` would be fine.

If there are binding attributes or `IWebJobsStartup` implementations in an assembly that doesn't follow the `Microsoft.Azure.WebJobs.Extensions.<name>` convention, we'd need to adjust this.

soninaren replied:
You are right, I saw the startup class implementation `DurableTask.SqlServer.AzureFunctions.dll` i assumed it to be implementing IWebJobsStartup , it is implementing 
 FunctionsStartup instead.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            var methodDecl = (MethodDeclarationSyntax)context.Node;
+            var methodName = methodDecl.Identifier.ValueText;
+
+            if (!HasFunctionNameAttribute(context, methodDecl))
```
Do we need the if condition and return block? Calling `HasFunctionNameAttribute` is good enough correct? I also think we should rename that method (HasFunctionNameAttribute) to be more explicit that it checks function name and report.

kashimiz replied:
Addressed in later comment.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            }
+
+            var methodDecl = (MethodDeclarationSyntax)context.Node;
+            var methodName = methodDecl.Identifier.ValueText;
```
nit: remove unused.

kashimiz replied:
Removed.

soninaren commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+
+        JobHostMetadataProvider _tooling;
+
+        internal JobHostMetadataProvider Tooling => _tooling;
```
Not sure why this is named tooling, Should we call it `metadataProvider or jobHostMetadataProvider `

kashimiz replied:
`jobHostMetadataProvider` it is.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            {
+                foreach (AttributeSyntax attributeSyntax in attrListSyntax.Attributes)
+                {
+                    // Perf - Can we get the name without doing a symbol resolution?
```
Would `GetText` method give us this info(name) ?

kashimiz replied:
`SyntaxNode.GetText` would give a string like `FunctionName("MyFunction")` that the function name could be extracted from. This approach would lose the symbol information, but we're only using the `NamedType.Name` string for comparison. It might be a bit fragile, though.

kshyju replied:
Got it. Thanks!

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+
+        // Does the method have a [FunctionName] attribute?
+        // This provides a quick check before we get into the more intensive analysis work.
+        private bool HasFunctionNameAttribute(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDeclarationSyntax)
```
IMHO, `Validate/CheckMethodsAndReport` or something along that line seems more readable (or something where we say we are reporting as well).

kashimiz replied:
Given that this isn't a simple boolean check and has some side effects, I think that's a good idea.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+                    {
+                        var attrType = symAttributeCtor.ContainingType;
+                        if (attrType.Name == nameof(FunctionNameAttribute))
+                        {
```
Also, consider inverting `if` blocks to reduce nesting. Helps with readability.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+                    }
+                }
+            }
+            return false;
```
Remove return value as those seems to be unused on the caller site.

kashimiz replied:
Good catch. I'm guessing this was implemented with a return value for flexibility, but ended up only being called in one place.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+        }
+
+        // Does the method have a [FunctionName] attribute?
+        // This provides a quick check before we get into the more intensive analysis work.
```
nit: Are these comments reflecting the latest version of the method?

kashimiz replied:
To the best of my knowledge, it did. I'm updating it to reflect the changes I'm pushing.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+                return;
+            }
+
+            // TODO: Attribute validation
```
nit: Cleanup if these are not needed anymore.

kashimiz replied:
That was addressed in this PR, so I'll take that comment out: https://github.com/Azure/azure-functions-host/pull/8400/

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,141 @@
+            var compilation = context.Compilation;
+
+            AssemblyCache.Instance.Build(compilation);
+            _tooling = AssemblyCache.Instance.Tooling;
```
Could we document what are the intentions of this method/why we are doing what we are doing here, to help future readers ? Also, for a present reader😁 I am lost here 😕

kashimiz replied:
Removed `AssemblyCache` for now after conversation offline; will reintroduce in the PR for the analyzer that actually makes use of it.

kshyju commented on src/WebJobs.Script.Analyzers/AssemblyCache.cs:
```diff
@@ -0,0 +1,190 @@
+
+namespace Microsoft.Azure.Functions.Analyzers
+{
+    // Only support extensions and WebJobs core.
```
Could we add some summary comment to this type explaining what this class does?

kashimiz replied:
Removed for now, but when it comes back it'll be with comments. :)

kshyju commented on src/WebJobs.Script.Analyzers/CompilationWebJobsStartuptypeLocator.cs:
```diff
@@ -0,0 +1,37 @@
+            return _startupTypes.Value;
+        }
+
+        public Type[] GetTypes()
```
Does this need to be public?

kashimiz replied:
No, no it does not.

kshyju commented on src/WebJobs.Script.Analyzers/DiagnosticDescriptors.cs:
```diff
@@ -27,5 +27,11 @@ private static DiagnosticDescriptor Create(string id, string title, string messa
                 severity: DiagnosticSeverity.Warning);
+
+        public static DiagnosticDescriptor IllegalFunctionName { get; }
+            = Create(id: "AZF0003", title: "Illegal Function name",
```
Which repo needs to be updated to make the https://docs.microsoft.com/azure/azure-functions/errors-diagnostics/sdk-rules/AZF0003 link working?

kashimiz replied:
There's an additional PR in [MicrosoftDocs/azure-docs](https://github.com/MicrosoftDocs/azure-docs) that I'll be linking.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobs.Script.Analyzers.csproj:
```diff
@@ -41,4 +45,7 @@
     </ItemGroup>
   </Target>
 
+  <Target Name="_AddNuGetDlls">
```
Did you mean to have this empty target?

kashimiz replied:
Ah, thanks. This is a remnant; found a cleaner way to include the WebJobs DLLs in the analyzer nuget. Removing.

liliankasem commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,126 @@
+        public static void VerifyWebJobsLoaded()
+        {
+            // Check if WebJobs types can be loaded
+            var jobHost = new JobHost(new OptionsWrapper<JobHostOptions>(new JobHostOptions()), null);
```
If we're not going to use the variable
```suggestion
            _ = new JobHost(new OptionsWrapper<JobHostOptions>(new JobHostOptions()), null);
```

kshyju replied:
I do not think this `VerifyWebJobsLoaded` method is needed in this analyzer PR anymore. Based on our last sync, we decided to keep only the code needed for the analyzer for function name validation. Feel free to correct me @kashimiz

kashimiz replied:
True; removed.

brettsam commented on src/WebJobs.Script.Analyzers/DiagnosticDescriptors.cs:
```diff
@@ -27,5 +27,11 @@ private static DiagnosticDescriptor Create(string id, string title, string messa
+
+        public static DiagnosticDescriptor IllegalFunctionName { get; }
+            = Create(id: "AZF0003", title: "Illegal Function name",
+                messageFormat: "Function name can't be '{0}'",
```
Is there any reasonable way for us to explain the regex to provide clues? Like "cannot be longer than 32 characters and must be alphanumeric?"

kashimiz replied:
How about: `Function name can't be '{0}'. It must start with a letter, be no longer than 128 characters, and contain only alphanumeric characters, underscores, or hyphens.`

128-character limit: https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs/FunctionNameAttribute.cs#L35

What logic imposes a 32-character limit at runtime? AFAIK we do this for host IDs, but I wasn't aware we imposed the same limit for function names.

brettsam replied:
That sounds good to me!

The 32-character limit is for host ids b/c we use that to create queues for some internal stuff, which have stricter limits on length. That logic is here -- https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Host/ScriptHostIdProvider.cs#L87

liliankasem reviewed (approved):
lgtm! just one last comment from me!

brettsam commented on test/WebJobs.Script.Tests.Analyzers/AssemblyCacheTests.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
delete this file for now since it's empty?

kashimiz replied:
Deleting.

brettsam commented on test/WebJobs.Script.Tests.Analyzers/WebJobs.Script.Tests.Analyzers.csproj:
```diff
@@ -7,6 +7,8 @@
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.WebJobs" Version="3.0.33" />
+    <PackageReference Include="Microsoft.Azure.WebJobs.Host.Storage" Version="4.0.4" />
```
do we need this storage ref here?

kashimiz replied:
Not for this analyzer, no. Thanks for the catch.

brettsam reviewed (commented):
Couple minor comments...

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,126 @@
+            context.RegisterSyntaxNodeAction(AnalyzeMethodDeclarationNode, SyntaxKind.MethodDeclaration);
+
+            // Hook compilation to get the assemblies' references and build the WebJob tooling interfaces.
+            context.RegisterCompilationStartAction(AnalyzeCompilation);
```
Do not think we need this one either in this analyzer PR as this is not really needed for function name validation analyzer.

kashimiz replied:
True; this analyzer doesn't use the tooling object. Thanks for the catch.

kshyju commented on src/WebJobs.Script.Analyzers/CompilationWebJobsStartuptypeLocator.cs:
```diff
@@ -0,0 +1,44 @@
+    /// Note: Assemblies are derived from the Compilation by AssemblyCache. 
+    /// This class pulls the WebJobsStartupAttributes from assemblies in the list.
+    /// </summary>
+    public class CompilationWebJobsStartupTypeLocator : IWebJobsStartupTypeLocator
```
Can delete this file as we do not need this file for the function name analyzer (current PR we want to merge). Add in the next PR where this will be actually used, as needed.

kashimiz replied:
Missed this; removing.

kshyju commented on src/WebJobs.Script.Analyzers/Constants.cs:
```diff
@@ -5,15 +5,22 @@ namespace Microsoft.Azure.Functions.Analyzers
 {
     internal static class Constants
     {
+        internal static class Assemblies
```
nit: Can name the class as `AssemblyNames` and const as `WebJobs`. So it will be used as `AssemblyNames.WebJobs`

kashimiz replied:
This is less redundant; I like it.

kshyju commented on src/WebJobs.Script.Analyzers/Constants.cs:
```diff
@@ -5,15 +5,22 @@ namespace Microsoft.Azure.Functions.Analyzers
         {
-            public const string Reliability = "Reliability";
-            public const string Usage = "Usage";
+            public const string Reliability = nameof(Reliability);
```
nit question: it's cool to know we can use nameof here. But If someone decides to do a rename on the caller site (by doing Ctrl + R +R ) on the constant name, will it also update the value?

kashimiz replied:
It would. "Which shouldn't be a problem because the tests would break," I thought, then looked at the tests, and it turns out they don't check expected vs actual `Category`. (There's not a way to do so without adding a custom `DiagnosticVerifier` either.)

tl;dr It's cool but you make a good point. Reverting to strings.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,86 @@
+                        // Validate the FunctionName
+                        var functionName = GetFunctionNameFromAttribute(context.SemanticModel, attributeSyntax);
+
+                        bool match = FunctionNameAttribute.FunctionNameValidationRegex.IsMatch(functionName);
```
The `IsMatch` method can throw when the input is `null`. `GetFunctionNameFromAttribute` have some code path to return `null`. So we should do a null check before calling the Match method.

kashimiz replied:
Good catch; thank you. Adding a null check.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,86 @@
+                foreach (AttributeSyntax attributeSyntax in attrListSyntax.Attributes)
+                {
+                    // Perf - Can we get the name without doing a symbol resolution?
+                    var test = attributeSyntax.GetText();
```
Remove this line. The variable `test` seems to be unused.

kashimiz replied:
Removed.

kshyju commented on src/WebJobs.Script.Analyzers/WebJobsAttributeAnalyzer.cs:
```diff
@@ -0,0 +1,86 @@
+
+        public override void Initialize(AnalysisContext context)
+        {
+            // https://stackoverflow.com/questions/62638455/analyzer-with-code-fix-project-template-is-broken
```
nit: Do we need to keep this comment?

kashimiz replied:
I would prefer to; it might save someone time/frustration in the future. (When I picked up analyzer work again after a break, I forgot to add the line below and lost a chunk of time figuring out why nothing was working. I have every reason to believe this will happen again. Probably to me.)

kshyju commented on src/WebJobs.Script.Analyzers/WebJobs.Script.Analyzers.csproj:
```diff
@@ -25,6 +25,9 @@
     <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="2.9.8" />
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.3.1" />
+	  
+	<!-- Packed in Microsoft.Azure.Functions.Analyzers NuGet -->
```
Do we really need this package reference? I wonder we should keep a copy of the Regex in the analyzer so that the analyzer won't have a dependency on WebJobs package. (I don't have a strong opinion though).

kashimiz replied:
The regex is unlikely to change, but I'd rather not give it the opportunity to get out of sync with what's in the WebJobs SDK. We'll need this reference later for the analyzer that runs through the binding rules (https://github.com/Azure/azure-functions-host/issues/8255) in any case.

kshyju commented on test/WebJobs.Script.Tests.Analyzers/WebJobsAttributeAnalyzerTests.cs:
```diff
@@ -0,0 +1,88 @@
+
+            var test = new AnalyzerTest();
+            test.ReferenceAssemblies = ReferenceAssemblies.Net.Net50.WithPackages(ImmutableArray.Create(
+                new PackageIdentity("Microsoft.NET.Sdk.Functions", "3.0.11"),
```
3.1.1 and 4.1.1 are the latest versions of `Microsoft.NET.Sdk.Functions` package.

kashimiz replied:
The other analyzers' tests need this change as well. I'll update this reference, but keep it as a separate commit when merging this PR.

kshyju reviewed (approved):
LGTM.

kashimiz committed:
Updated analyzer tests' reference assemblies:
- Microsoft.NET.Sdk.Functions 3.0.11 --> 3.1.1
- Microsoft.Azure.WebJobs.Extensions.Storage 3.0.10 --> 3.0.11
